Modules/_io/textio.c

```diff
@@ -1060,21 +1060,31 @@ _io_TextIOWrapper___init___impl(textio *self, PyObject *buffer,
 
     self->has_read1 = _PyObject_HasAttrId(buffer, &PyId_read1);
 
-    self->encoding_start_of_stream = 0;
     if (self->seekable && self->encoder) {
         PyObject *cookieObj;
         int cmp;
 
         self->encoding_start_of_stream = 1;
 
-        cookieObj = PyObject_CallMethodObjArgs(buffer, _PyIO_str_tell, NULL);
-        if (cookieObj == NULL)
-            goto error;
+        if (Py_TYPE(buffer) == &PyBufferedWriter_Type
+           || Py_TYPE(buffer) == &PyBufferedRandom_Type) {
+            /* Fast-path for _io.BufferedWriter and _io.BufferedRandom:
```

Comment:
we only go into this path if self->encoder is set. self->encoder is only set if the buffer is writable.